### PR TITLE
Expose some homeserver functionality to spam checkers

### PR DESCRIPTION
**Pull request by turt2live**
Primarily for use in https://github.com/matrix-org/mjolnir/pull/7

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off) (NV hat)

### CODE REVIEW

File: synapse/events/spamcheck.py [hunk 1]
```diff
@@ -26,7 +29,13 @@ def __init__(self, hs):
             pass
 
         if module is not None:
-            self.spam_checker = module(config=config)
+            # Older spam checkers don't accept the `hs` argument, so we
+            # try and detect support.
+            spam_args = inspect.getargspec(module)
```
Comment (auscompgeek) [inline synapse/events/spamcheck.py]: You'll want to use either `getfullargspec` or `signature` instead, as `getargspec` will bail if any function annotations exist.

**Comment by erikjohnston** [comment]
I've tried to avoid passing homeserver object to the modules, as it basically ends up forcing us to keep the entire Synapse API stable so as to not break installed modules. Generally we've tried to document the interface of the objects we're passing in, e.g. for password providers we pass in https://github.com/matrix-org/synapse/blob/fa7cf8778eccee209b8ffadb71bbe33d12fa0f97/synapse/module_api/__init__.py#L24

**Comment by turt2live** [comment]
@erikjohnston please take another look

**Comment by turt2live** [comment]
Looks like the linter is failing on this, but for everything except the code I'm adding?

**Comment by erikjohnston** [comment]
> Looks like the linter is failing on this, but for everything except the code I'm adding?

The linters broke, should be fixed by merging in develop.

---

As per #synapse-dev:matrix.org, we've tried hard to avoid giving the homeserver object to pluggable modules, as otherwise its impossible to know what interfaces modules depend on. This adds a maintenance burden to synapse devs (makes it harder to know what functionality is relied on by third party modules), third party module devs (they don't know what APIs are stable or when they might just break) and server admins (don't know whether a new version of synapse will break their modules or not without testing, possibly causing service outages).

The flip side to this is that its more annoying to develop modules that use functionality that is not currently exposed by Synapse, as it means opening a PR on Synapse and waiting for that to be merged and released.

However given the fact that modules hook into core parts of Synapse and are security sensitive (e.g. auth providers), we think that the need for stability outweighs the need to be able to quickly add new features.

As such the solution we've adopted is to pass in a well-defined [ModuleApi](https://github.com/matrix-org/synapse/blob/782dd72037cf71fb3f9e4922b07c56df2f59de75/synapse/module_api/__init__.py#L24) interface to modules, and I'd suggest you use that and adding any necessary extra functionality. (Other potential solutions exist, including adding version constraints, but none of those have been implemented or thought through.)

**Comment by turt2live** [comment]
I added said module the day we had that discussion - please take a look.

**Comment by erikjohnston** [comment]
Argh, sorry Travis I completely missed that this had been updated :man_facepalming:

**Pull request merged**